src/runtime/cuda.cpp

```diff
@@ -109,6 +109,13 @@ WEAK int halide_cuda_acquire_context(void *user_context, CUcontext *ctx, bool cr
             __sync_lock_release(&thread_lock);
             return error;
         }
+        // Creation automatically pushes the context, but we'll pop so the caller
```

Comment:
Yeah, I also think it should go in create_cuda_context, though I don't feel
strongly about it.

I also agree that your suggested design is a little cleaner. Currently the
logic is spread out across the Context class, the acquire/release
functions, and the device release function.

On Sun, Nov 8, 2015 at 11:17 PM, Zalman Stern <notifications@github.com>
wrote:

> In src/runtime/cuda.cpp
> <https://github.com/halide/Halide/pull/964#discussion_r44246275>:
>
> > @@ -109,6 +109,13 @@ WEAK int halide_cuda_acquire_context(void *user_context, CUcontext *ctx, bool cr
> >              __sync_lock_release(&thread_lock);
> >              return error;
> >          }
> > +        // Creation automatically pushes the context, but we'll pop so the caller
>
> I would prefer this code to go in create_cuda_context so that it sits
> right next to the cuCtxCreate call.
>
> Thinking about this, I'll toss out what may be a better design for others,
> specifically, Andrew and Dillon, to comment on. The idea is to move the
> push/pop behavior into
> halide_cuda_acquire_context/halide_cuda_release_context and make it clear
> these are always called in bracketed pairs on a single thread. That way,
> inside halide_cuda_acquire_context either cuCtxCreate or cuCtxPushCurrent
> would be called, but not both.
>
> Part of the reason for this is it seems to me halide_cuda_device_release
> is depending on the context being made current even though it does not use
> Context. I would also add a "destroy" argument to
> halide_cuda_release_context, which would a) allow those overriding the
> acquire/release mechanism to get notification when the context can be freed
> as it is unlikely to be reused, b) would get rid of the code that tests to
> ctx == context to decide to call cuCtxDestroy, and c) would mirror the
> create argument to halide_cuda_acquire_context.
>
> This is a small change to a public facing API in that anyone who has
> overridden halide_cuda_acquire_context/halide_cuda_release_context now has
> to handle the push/pop behavior and handle destruction in
> halide_cuda_release_context, but it seems like a cleaner design. Of course
> we should probably also mirror this in OpenCL.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/halide/Halide/pull/964/files#r44246275>.
>